### PR TITLE
error handling in topology local

### PR DESCRIPTION
**Pull request by rupnikj**

### CODE REVIEW

File: src/topology_local.ts [hunk 1]
```diff
@@ -173,32 +172,33 @@ export class TopologyLocal {
...
-            this.isRunning = false;
-        } catch (e) {
-            callback(e);
+        if (!this.isInitialized) { 
+            return callback(new Error(this.logging_prefix + "Topology not initialized and cannot be paused."));
         }
-        callback();
+        // spouts pass internal exceptions to errorCallback
+        for (let spout of this.spouts) {
+            spout.pause();
```
Comment (bergloman) [inline src/topology_local.ts]: what if `spout.pause()` raises an exception? you removed try/catch

File: src/topology_local.ts [hunk 2]
```diff
@@ -208,6 +208,11 @@ export class TopologyLocal {
             clearInterval(self.heartbeatTimer);
         }
         self.pause((err) => {
+            if (err) {
+                // only possible error is when isInit is false
+                log.logger().error("THIS SHOULD NOT HAPPEN!");
+                log.logger().exception(err);
+            }
```
Comment (bergloman) [inline src/topology_local.ts]: why don't we call `this.onInternalError()` here?

File: src/topology_local.ts [hunk 3]
```diff
@@ -274,13 +279,17 @@ export class TopologyLocal {
             try {
                 spout.heartbeat();
             } catch (e) {
+                // All exceptions should have been caught and passed to errorCallback
+                log.logger().error("THIS SHOULD NOT HAPPEN!");
```
Comment (bergloman) [inline src/topology_local.ts]: why don't we call `this.onInternalError()` here?

File: src/topology_local.ts [hunk 4]
```diff
@@ -274,13 +279,17 @@ export class TopologyLocal {
...
                 log.logger().exception(e);
             }
         }
         for (let bolt of this.bolts) {
             try {
                 bolt.heartbeat();
             } catch (e) {
+                // All exceptions should have been caught and passed to errorCallback
+                log.logger().error("THIS SHOULD NOT HAPPEN!");
```
Comment (bergloman) [inline src/topology_local.ts]: why don't we call `this.onInternalError()` here?

File: src/topology_local.ts (hunk 1 above)
Comment (rupnikj) [reply src/topology_local.ts]: Because everything in spout.pause is in try - catch

Comment (bergloman) [reply src/topology_local.ts]: I don't see that :)
https://github.com/rupnikj/qtopology/blob/f7342b845a9b4c077339c1f370c1f69488a7c9bb/src/topology_local.ts#L183

what am I missing?

Comment (rupnikj) [reply src/topology_local.ts]: https://github.com/rupnikj/qtopology/blob/master/src/topology_local_inprocess.ts#L272
each spout.pause cannot throw

**Pull request merged**